internal/addrs/partial_expanded.go

```diff
@@ -258,6 +413,99 @@ type PartialExpandedResource struct {
 	resource Resource
 }
 
+// ParsePartialExpandedResource parses a resource address traversal and returns
+// a PartialExpandedResource representing the known and unknown parts of the
+// address.
+func ParsePartialExpandedResource(traversal hcl.Traversal) (PartialExpandedResource, hcl.Traversal, tfdiags.Diagnostics) {
+	var diags tfdiags.Diagnostics
+
+	pem, remain, diags := ParsePartialExpandedModule(traversal)
+	if len(remain) == 0 {
+		diags = diags.Append(&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Invalid address",
+			Detail:   "Resource address must be a module address followed by a resource address.",
+			Subject:  traversal.SourceRange().Ptr(),
+		})
+		return PartialExpandedResource{}, nil, diags
+	}
+
+	mode := ManagedResourceMode
+	if remain.RootName() == "data" {
+		mode = DataResourceMode
+		remain = remain[1:]
+	}
+
+	if len(remain) < 2 {
+		diags = diags.Append(&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Invalid address",
+			Detail:   "Resource specification must include a resource type and name.",
+			Subject:  remain.SourceRange().Ptr(),
+		})
+		return PartialExpandedResource{}, nil, diags
+	}
+
+	var typeName, name string
+	switch tt := remain[0].(type) {
+	case hcl.TraverseRoot:
```

Comment:
Above should ensure that this is a `TraverseRoot`, and would also panic with `RootName()` if it weren't.

Reply:
Thanks. Now I see a key point we need to address, managed resources and data resources shouldn't be handled that differently. A managed resource can be prefixed with `resource.` just like a data source is prefixed with `data.`. I just checked though and while reference parsing works, the target addr parsing code was never updated, but we should probably start taking this into account to clean things up though.

Reply:
I'll put a follow up PR together that fixes this both here and in the target parsing function.